### ISSUE TITLE
Add info-only option for enabling checks

### ISSUE THREAD

**Issue by avylove** [PyCQA/pylint#3795]
### Is your feature request related to a problem? Please describe
Currently individual checks can be enabled or disabled. There is additional controls over how the level/category of a check affects the score/return code, but there is no capability to enable a check as only informational.

### Describe the solution you'd like
Add a new CLI and config option "info-only" which will provide normal output of a failed check, but the check failure will not affect the score/return code. The failure message should also indicate the check is info-only.

### Additional context
This will greatly improve the workflow for enabling checks on a large codebase. Currently checks need to be globally disabled or disabled on a per file basis until the issues are resolved. Sometimes this can take a long time and having output to advise developers of checks that you aren't currently failing against will encourage them to resolve those issues when touching related code.

**Comment by KebabLord** [comment]
agree

**Comment by papb** [comment]
Maybe #3251 could also solve this

**Comment by Pierre-Sassoulas** [comment]
Closing as I think there's better solutions:
- Long term with https://github.com/PyCQA/pylint/issues/5604
- Short term by launching  pylint with more checks  in an optional job in continuous integration OR locally when you actually want to read what pylint have to say (the latter will save you a **lot** of server time and noise that no one ever read in the continuous integration)

**Comment by avylove** [comment]
@Pierre-Sassoulas , I think the use case here is different. The case #5604 seems to be addressing is applying pylint to a codebase that isn't already utilizing pylint. While info-only messages can definitely help in that use case, the use case here is a codebase that is already using pylint and either wants to enable checks that were previously disabled or upgrade pylint without fixing new checks right away. On a large codebase it may take some time to work through all the failures and you don't want that to be blocking, but it is still valuable to report the failures so developers see them as they are introducing changes.

**Comment by jacobtylerwalls** [comment]
I think Pierre's second proposal is still useful -- just have an optional job that doesn't fail and doesn't emit low scores.

We let folks control their score calculations (and thus exit codes given `--fail-under`) with the `--evaluation` setting. Maybe we could provide a new variable for use there like `custom` that reads from an option `--apply-custom-scoring=`, and then you can calculate the score however you like.

**Comment by Pierre-Sassoulas** [comment]
> applying pylint to a codebase that isn't already utilizing pylint

For that use case our default value should be reasonable enough so that you don't have to enable/disable them and following pylint message should keep the code "under control". So this is not what #5604 is about, I want to make it easier to increase the code quality with a pylint "guide". Quoting #5604:

> The first step is Implementing baseline so only new message are raised : https://github.com/PyCQA/pylint/issues/5403.

So only the new message need to be fixed and not all the legacy one whatever the configuration you use.

> upgrading the pylintrc step by step to fix the legacy messages starting by the more significant one (error, then vicious warning, then warning, then important convention, then opinionated convention, then adding extensions...).

The idea is that pylint should be able to tell you what message is the most important to fix instead of dumping every message on you. Based on what is ignored in the baseline or based on what is currently disabled in your configuration. 

Please feel free to contribute to the conversation on #5604, it did not gather a lot yet and it makes me sad :)

**Comment by avylove** [comment]
That's what I'm saying. #5604 seems like making pylint more like the way coveralls deltas work. I see a value in that, but that is not my use case. The current capabilities also do not seem to provide what I need either.

Currently there seems to be two ways checks are handled: enabled and disabled. I'm asking for a third, enabled-informationally. I suspect for #5604 you'll want another to distinguish checks that are always required to pass from checks that can fail if the score isn't reduced.

**Comment by jacobtylerwalls** [comment]
@avylove do you have a reaction to my suggestion of a new variable `custom` for use in calculating the score that could read from `--apply-custom-scoring=`?

That would be a small change suitable for a new pylint contributor and it would achieve "informational messages" by zeroing out their influence on the score, right?

**Comment by avylove** [comment]
@jacobtylerwalls You'd still need a way to bucket the checks, so it sounds like what you're talking about is a plugin interface allowing someone to write a plugin with additional config options specifying a zero score for some checks. Technically that could work but you may have to list the same check twice in the config. The other issue is you'd still need a way to inform the user the checks were informational only and are not being score. I'm not sure how that would be handled in this implementation.

**Comment by jacobtylerwalls** [comment]
Just to clarify, I'm not talking about a plugin, I'm talking about a command-line (or config) option that puts the messages into a custom score (default zero effect). Something like:

```shell
pylint mymodule --custom-scoring=missing-module-docstring
```
```shell
************* Module mymodule
mymodule.py:1:0: C0114: Missing module docstring (missing-module-docstring)

------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 0.00/10, +10.00)
```

I know I'm proposing a bit less than you b/c there's no feedback about which checks were informational only.

**Comment by Pierre-Sassoulas** [comment]
The underlying issue that is trying to be fixed here as far as I understand is being able to make the codebase better little by little. We have two long term solutions already with #5403 and #5604. The enable/disable and options' constellation around it is complex enough as it is I'm opposed to add layer of complexity around it.

What I recommend until those two issues are implemented in pylint is to launch pylint with ``--enable=all --enable-all-extension`` (or a list of messages you identified) **when you actually want to read pylint's output**. Adding informational warning is going to cause **[alarm fatigue](https://en.wikipedia.org/wiki/Alarm_fatigue) and those informational warning are going to be ignored by user 95+% of the time. It's also going to make pylint's runs take longer.** 

But if you really insist on shooting yourself in the foot you can use ``--fail-on=F,E,C0111, --exit-zero`` with ``F,E,C0111`` the list of check you want to actually fail (see ``pylint --help`` or the [documentation](https://pylint.pycqa.org/en/latest/user_guide/configuration/all-options.html?highlight=%22%20--fail-on%22#fail-on)).

**Comment by avylove** [comment]
@Pierre-Sassoulas, I apologize if I failed to communicate the ask effectively. It is not to make the codebase better little by little. The assumption here is you already have Pylint enabled and configured. Then you upgrade pylint or enable and extension and introduce a new check (basic check maintenance) that will take multiple PRs to work through, so before you begin failing on a check, you want to warn users when the condition is met.

This request came about because I was working on a large codebase and when introducing new checks I found as I made fixes other devs would commit code causing the same issue I was attempting to fix. This resulted in multiple "cleanup" PRs or a broken build pipeline before I was able to get a new check enabled. For other precommit checks we were able to give informational messages of upcoming changes and this caused the code reviewer to delay approval until the issues were resolved even though the CI system did not fail the build.

Alarm fatigue assumes there are constantly warning messages, which is not the case here. This is intended to be a temporary measure of ongoing check maintenance.

I don't see how that can be accomplished with what you suggested. What you seem to be suggesting is running pylint twice, once with informational-only checks disabled and once with them enabled and ignore the result.

**Comment by Pierre-Sassoulas** [comment]
> This request came about because I was working on a large codebase and when introducing new checks I found as I made fixes other devs would commit code causing the same issue I was attempting to fix. This resulted in multiple "cleanup" PRs or a broken build pipeline before I was able to get a new check enabled. 

It's currently possible to add [inline or per file disable in your code](https://pylint.pycqa.org/en/latest/user_guide/messages/message_control.html) to disable existing violations, which seems to achieve what you want. #5403 would create a way to hide all existing violations without having to disable them one by one, so the same thing but with zero work to do on your side.

> What you seem to be suggesting is running pylint twice, once with informational-only checks disabled and once with them enabled and ignore the result.

Well yes, it's the thing that would require zero work to put into place. It's also possible to have message that won't error ``--fail-on=F,E,C0111, --exit-zero`` but you'll have to do a list of the message that you want to fail in fail-on.

**Comment by avylove** [comment]
> It's currently possible to add inline or per file disable in your code to disable existing violations, which seems to achieve what you want.

I specifically call this out in the original request. This works in some cases, but does not scale for this use case.

> It's also possible to have message that won't error --fail-on=F,E,C0111, --exit-zero but you'll have to do a list of the message that you want to fail in fail-on.

This doesn't inform the user which checks are informational and which are not.

As I've stated multiple times, #5403 is trying to solve a different problem. That said, if #5043 is implemented as a CI service, then maybe that will allow the level of flexibility needed, but that would be external to pylint and may or may not have a self-hosting option.

I think I've failed to convey to need. Seems like to only current viable option is to run pylint twice, which seems wasteful.

**Comment by avylove** [comment]
@DanielNoord Yes, that's it exactly! And yes, an option like `--tell-me-about-this-but-don't-fail` would work. There would just need to be some indication in the output to differentiate these from failures that do cause the CI to fail.

**Comment by jacobtylerwalls** [comment]
> There would just need to be some indication in the output to differentiate these from failures that do cause the CI to fail.

How about:
```python
************* Module mymodule
mymodule.py:1:0: C0114: Missing module docstring (missing-module-docstring)

------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 0.00/10, +10.00)
Custom scoring applied to: missing-module-docstring
```

I don't want `pylint` to become responsible for knowing how you've rigged your CI step to fail or pass, which is controlled by `--fail-on`, `--fail-under`, `--exit-zero`, and `--evaluation` and now `--custom-scoring` before you even exit the pylint layer. But I could see adding one line to the output like the above. Still, I'm not really sure why it's needed -- isn't the goal to get the developers to fix the issues whether they're required by CI or not? Is it that developers will be confused why the workflow step is green if they see output?

**Comment by avylove** [comment]
@jacobtylerwalls  `--custom-scoring` seems like you would point to a subclass of a scoring class which implements completely different scoring logic for everything.

For `--tell-me-about-this-but-don't-fail` (or maybe `--enable-info-only`), I think it would need to be a sublist or include an indicator. Something like:

```
************* Module mymodule
mymodule.py:6:0: C0305: Trailing newlines (trailing-newlines)
------------- Informational only
mymodule.py:1:0: C0114: Missing module docstring (missing-module-docstring)

------------------------------------------------------------------
Your code has been rated at 9.00/10 (previous run: 0.00/10, +9.00)
```

or 

```
************* Module mymodule
mymodule.py:6:0: C0305: Trailing newlines (trailing-newlines)
(*) mymodule.py:1:0: C0114: Missing module docstring (missing-module-docstring)

------------------------------------------------------------------
Your code has been rated at 9.00/10 (previous run: 0.00/10, +9.00)
(*) Informational only
```
